language-server/test/dotty/tools/languageserver/util/CodeTester.scala

```diff
@@ -30,6 +33,32 @@ class CodeTester(projects: List[Project]) {
 
   private val positions: PositionContext = getPositions(files)
 
+  /** Check that the last diagnostics that have been published so far by the server
+   *  for a given file match `expected`.
+   *
+   * @param marker   The marker defining the source file from which to query.
+   * @param expected The expected diagnostics to be found
+   */
+  def diagnostics(marker: CodeMarker,
+      expected: (CodeRange, String, DiagnosticSeverity, Option[ErrorMessageID])*): this.type = {
+    implicit val posCtx = positions
+
+    def toDiagnostic(range: CodeRange, message: String, severity: DiagnosticSeverity,
+        errorCode: Option[ErrorMessageID]): Diagnostic = {
+      val location = range.toLocation
+      new Diagnostic(
+        location.getRange, message, severity, /*source=*/"", errorCode.map(_.errorNumber.toString).getOrElse(null)
+      )
+    }
+
+    val expectedParams = marker.toPublishDiagnosticsParams(expected.toList.map(toDiagnostic))
+    val actualParams = testServer.client.diagnostics.get.reverse.find(_.getUri == marker.uri)
+      .getOrElse(throw new Exception(s"No published diagnostics for ${marker.uri}"))
+    assertEquals(expectedParams, actualParams)
+
+    this
+  }
```

Comment:
The other "actions" in this file usually do
```scala
def foo(marker: CodeMarker, expected: ...) = doAction(new Foo(marker, expected))
```

Any reason for not following the same pattern?

Reply:
Mostly because there's no action (request sent to the server) involved here, the request is "didOpen" but it's done implicitly. But if you think this could be refactored, please do!